src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDiagnosticFactory.cs

```diff
@@ -421,5 +421,48 @@ public static RazorDiagnostic Create_RefSuppressFieldRequiredForGeneric(SourceSp
         {
             return RazorDiagnostic.Create(RefSuppressFieldRequiredForGeneric, source ?? SourceSpan.Undefined);
         }
+
+        public static readonly RazorDiagnosticDescriptor ComponentNamesCannotStartWithLowerCase =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10013",
+                () => "Component '{0}' starts with a lowercase character. Component names cannot start with a lowercase character.",
+                RazorDiagnosticSeverity.Error);
+
+        public static RazorDiagnostic Create_ComponentNamesCannotStartWithLowerCase(string componentName, SourceSpan? source = null)
+        {
+            return RazorDiagnostic.Create(
+                ComponentNamesCannotStartWithLowerCase,
+                source ?? SourceSpan.Undefined,
+                componentName);
+        }
+
+        public static readonly RazorDiagnosticDescriptor UnexpectedMarkupElement =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10014",
+                () => "Found markup element with unexpected name '{0}'. If this is intended to be a component, add a @using directive for its namespace.",
```

Comment:
What do they do if it isn't intended to be a component. Does using `!` help?

Reply:
Yes